akka-discovery-azure-api/src/main/scala/akka/discovery/azureapi/AzureRbacAKSServiceDiscovery.scala

```diff
@@ -0,0 +1,270 @@
+    AzureRbacAKSServiceDiscovery.azureDefaultCredential
+      .getToken(AzureRbacAKSServiceDiscovery.accessTokenRequestContext.addScopes(settings.entraServerId))
+      .toFuture
+      .asScala
+
+  private val kubernetesSetup = {
+    implicit val blockingDispatcher: ExecutionContext = system.dispatchers.lookup(DefaultBlockingDispatcherId)
+    for {
+      namespace: String <- Future {
+        settings.podNamespace
+          .orElse(readConfigVarFromFilesystem(settings.podNamespacePath, "pod-namespace"))
+          .getOrElse("default")
+      }
+      httpsContext <- Future(clientHttpsConnectionContext())
+    } yield {
+      KubernetesSetup(namespace, httpsContext)
+    }
+  }
+
+  import system.dispatcher
+
+  private def accessToken(retries: Int = 3): Future[AccessToken] = {
+    fetchToken.flatMap { token =>
+      if (token.isExpired && retries > 0) {
+        accessToken(retries - 1)
+      } else if (token.isExpired) {
+        Future.failed(new RuntimeException("Failed to fetch a valid token after multiple attempts"))
+      } else {
+        Future.successful(token)
+      }
+    }
+  }
+
+  override def lookup(lookup: Lookup, resolveTimeout: FiniteDuration): Future[ServiceDiscovery.Resolved] = {
+    val selector = settings.podLabelSelector(lookup.serviceName)
+
+    for {
+      ks <- kubernetesSetup
+
+      at <- accessToken()
+
+      request <- {
+        log.info(
+          "Querying for pods with label selector: [{}]. Namespace: [{}]. Port: [{}]",
+          selector,
+          ks.namespace,
+          lookup.portName
+        )
+
+        optionToFuture(
+          // FIXME | remove .get
+          podRequest(at.getToken, ks.namespace, selector),
+          s"Unable to form request; check Kubernetes environment (expecting env vars ${settings.apiServiceHostEnvName}, ${settings.apiServicePortEnvName})"
+        )
+      }
+
+      response <- http.singleRequest(request, ks.ctx)
+
+      entity <- response.entity.toStrict(resolveTimeout)
+
+      podList <- {
+
+        response.status match {
+          case StatusCodes.OK =>
+            log.debug("Kubernetes API entity: [{}]", entity.data.utf8String)
+            val unmarshalled = Unmarshal(entity).to[PodList]
+            unmarshalled.failed.foreach { t =>
+              log.warning(
+                "Failed to unmarshal Kubernetes API response.  Status code: [{}]; Response body: [{}]. Ex: [{}]",
+                response.status.value,
+                entity,
+                t.getMessage)
+            }
+            unmarshalled
+          case StatusCodes.Forbidden =>
+            Unmarshal(entity).to[String].foreach { body =>
+              log.warning(
+                "Forbidden to communicate with Kubernetes API server; check RBAC settings. Response: [{}]",
+                body)
+            }
+            Future.failed(
+              new KubernetesApiException("Forbidden when communicating with the Kubernetes API. Check RBAC settings."))
```

Comment:
Hum, this means the warning could is logged independently of returning the failure, before or after. Could perhaps be worth to make sure it is sequential for more predictability (`Future#andThen`)